source/common/config/watch_map.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include <set>
+#include <string>
+#include <utility>
+
+#include "envoy/config/subscription.h"
+
+#include "common/common/assert.h"
+#include "common/common/logger.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+
+namespace Envoy {
+namespace Config {
+
+// Manages "watches" of xDS resources. Several xDS callers might ask for a subscription to the same
+// resource name "X". The xDS machinery must return to each their very own subscription to X.
+// The xDS machinery's "watch" concept accomplishes that, while avoiding parallel reduntant xDS
+// requests for X. Each of those subscriptions is viewed as a "watch" on X, while behind the scenes
+// there is just a single real subscription to that resource name.
+//
+// This class maintains the watches<-->subscription mapping: it
+// 1) delivers updates to all interested watches, and
+// 2) tracks which resource names should be {added to,removed from} the subscription when the
+//    {first,last} watch on a resource name is {added,removed}.
+//
+// #1 is accomplished by WatchMap's implementation of the SubscriptionCallbacks interface.
+// This interface allows the xDS client to just throw each xDS update message it receives directly
+// into WatchMap::onConfigUpdate, rather than having to track the various watches' callbacks.
+//
+// The information for #2 is returned by updateWatchInterest(); the caller should use it to
+// update the subscription accordingly.
+//
+// A WatchMap is assumed to be dedicated to a single type_url type of resource (EDS, CDS, etc).
+class WatchMap : public SubscriptionCallbacks, public Logger::Loggable<Logger::Id::config> {
+public:
+  WatchMap() {}
+
+  // An opaque token given out to users of WatchMap, to identify a given watch.
+  using Token = uint64_t;
+
+  // Adds 'callbacks' to the WatchMap, with no resource names being watched.
+  // (Use updateWatchInterest() to add some names).
+  // Returns a new token identifying the newly added watch.
+  Token addWatch(SubscriptionCallbacks& callbacks);
+
+  // Returns true if this was the very last watch in the map.
+  // Expects that the watch to be removed has already had all of its resource names removed via
+  // updateWatchInterest().
+  bool removeWatch(Token token);
+
+  // Updates the set of resource names that the given watch should watch.
+  // Returns any resource name additions/removals that are unique across all watches. That is:
+  // 1) if 'resources' contains X and no other watch cares about X, X will be in pair.first.
+  // 2) if 'resources' does not contain Y, and this watch was the only one that cared about Y,
+  //    Y will be in pair.second.
+  std::pair<std::set<std::string>, std::set<std::string>>
```

Comment:
I considered replacing this with a struct with 'added' and 'removed' fields, to get something clearer than .first and .second. Not sure if the clarity is worth the complexity, what do you think?

Reply:
Done.